File: src/main/java/team4/footwithme/member/api/request/validation/PasswordMatchesValidator.java
package team4.footwithme.member.api.request.validation;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import java.lang.reflect.Field;

public class PasswordMatchesValidator implements ConstraintValidator<PasswordMatches, Object> {

    private String passwordField;
    private String passwordConfirmField;

    @Override
    public void initialize(PasswordMatches constraintAnnotation) {
        this.passwordField = constraintAnnotation.passwordField();
        this.passwordConfirmField = constraintAnnotation.passwordConfirmField();
    }

    @Override
    public boolean isValid(Object value, ConstraintValidatorContext constraintValidatorContext) {

        try {
            Field password = value.getClass().getDeclaredField(passwordField);
            Field passwordConfirm = value.getClass().getDeclaredField(passwordConfirmField);

            password.setAccessible(true);
            passwordConfirm.setAccessible(true);

            String passwordValue = (String) password.get(value);
            String passwordConfirmValue = (String) passwordConfirm.get(value);

            if(passwordValue == null && passwordConfirmValue == null){ // OAuth 2.0 로그인 일 시
                return true;
            }

            return passwordValue.equals(passwordConfirmValue);
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

}

File: src/main/java/team4/footwithme/member/api/request/validation/PasswordMatchesValidator.kt
package team4.footwithme.member.api.request.validation

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext

class PasswordMatchesValidator : ConstraintValidator<PasswordMatches, Any> {
    private var passwordField: String? = null
    private var passwordConfirmField: String? = null

    override fun initialize(constraintAnnotation: PasswordMatches) {
        this.passwordField = constraintAnnotation.passwordField
        this.passwordConfirmField = constraintAnnotation.passwordConfirmField
    }

    override fun isValid(value: Any, constraintValidatorContext: ConstraintValidatorContext): Boolean {
        try {
            val password = value.javaClass.getDeclaredField(passwordField)
            val passwordConfirm = value.javaClass.getDeclaredField(passwordConfirmField)

            password.isAccessible = true
            passwordConfirm.isAccessible = true

            val passwordValue = password[value] as String
            val passwordConfirmValue = passwordConfirm[value] as String

            if (passwordValue == null && passwordConfirmValue == null) { // OAuth 2.0 로그인 일 시
                return true
            }

            return passwordValue == passwordConfirmValue
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        }
    }
}